### PR TITLE
Support multiple handlers dealing with the same command

### PR DESCRIPTION
**Pull request by CumpsD**
In the original implementation trying to do the following was not possible:

```csharp
private class CommandHandlerModule1 : CommandHandlerModule
{
    public int CommandCounter;

    public CommandHandlerModule1()
    {
        For<Command>()
            .Handle((_, __) => Task.FromResult((long) CommandCounter++));
    }
}

private class CommandHandlerModule2 : CommandHandlerModule
{
    public int CommandCounter;

    public CommandHandlerModule2()
    {
        For<Command>()
            .Handle((_, __) => Task.FromResult((long) CommandCounter++));
    }
}
```

This would throw an `InvalidOperationException` saying ` Attempt to register multiple handlers for command type Command`.

This doesn't make sense. Separation wise I like to create multiple `CommandHandlerModule`s which do something completely different with a command.

I've changed to code to allow multiple `CommandHandlerModule`s to deal with the same command, while still remaining async. I added two tests as well to verify the correct behavior.

There is another check like this in the `CommandHandlerModule` itself which prevents multiple `For`'s to listen to the same. I'm fine with leaving this one since you are already in the context/mindset of the class.

```csharp
public virtual ICommandHandlerBuilder<CommandMessage<TCommand>> For<TCommand>()
    where TCommand : class
    => new CommandHandlerBuilder<TCommand>(handlerRegistration =>
    {
        if (!HandlerRegistrations.Add(handlerRegistration))
            throw new InvalidOperationException(
                "Attempt to register multiple handlers for command type {0}".FormatWith(typeof(TCommand)));
    });
```

I would make this into a breaking change because multiple handlers mean they can return multiple results (the `long`). While this is not used here, it is probably used upstream which will need to be changed to decide what to do with multiple results (pick the lowest/highest).

PS: About the _why_ a handler needs to return a `long`, that's not part for this PR :)

### CODE REVIEW

File: src/Be.Vlaanderen.Basisregisters.CommandHandling/CommandHandlerResolver.cs [hunk 1]
```diff
@@ -2,33 +2,30 @@ namespace Be.Vlaanderen.Basisregisters.CommandHandling
...
         {
             foreach(var module in commandHandlerModules)
             {
                 foreach(var handlerRegistration in module.HandlerRegistrations)
                 {
-                    if (!_knownCommandTypes.Add(handlerRegistration.CommandType))
-                        throw new InvalidOperationException(
-                            "Attempt to register multiple handlers for command type {0}".FormatWith(handlerRegistration.CommandType));
+                    if (!_handlers.ContainsKey(handlerRegistration.RegistrationType))
+                        _handlers[handlerRegistration.RegistrationType] = new List<object>();
```
Comment (ridingwolf) [inline src/Be.Vlaanderen.Basisregisters.CommandHandling/CommandHandlerResolver.cs:18]: This allows to register duplicates of the same handler, not just registering multile different handlers for the same command type.

Shouldn't there be a guard to check if that specific handler is already registered?

Comment (CumpsD) [reply src/Be.Vlaanderen.Basisregisters.CommandHandling/CommandHandlerResolver.cs:18]: If you register a duplicate, wouldn't the `if ContainsKey` prevent it from being registered twice?

**Comment by yreynhout** [comment]
If you want to do different things with the same command, why not put the delegation in the command handler itself, instead of pushing this down to infrastructure? This is just going to confuse the heck out of people trying to reason about the code because it's not obvious how this will get routed and the result combination, well, seem like accidental complexity IMO.

**Comment by CumpsD** [comment]
@yreynhout trying to accomodate for feature based command handler modules.

Fictive example:

```csharp
public class CachePopulator : CommandHandlerModule
{
    public CachePopulator()
    {
        For<CreateNewThing>()
            .Handle((_, __) => ...... populate a redis cache......);
    }
}

public class FtpDump : CommandHandlerModule
{
    public FtpDump()
    {
        For<CreateNewThing>()
            .Handle((_, __) => ...... create a file on some ftp......);
    }
}
```

Instead of:

```csharp
public class ThingHandler : CommandHandlerModule
{
    public FtpDump()
    {
        For<CreateNewThing>()
            .Handle((_, __) => 
            {
                ...... populate a redis cache.....
                ...... create a file on some ftp......
               ...15 more dispatches....
            });
    }
}
```

**Comment by CumpsD** [comment]
I don't think it should complicate reasoning about the code too much. Most use cases will remain the same as today, 1 module with handlers.

And if it's used, it's quite similar to EventHandlers, you also have multiple classes listening to the same events.

I could look into adding it as an overload to not be breaking. Then it's only the more advanced cases who need it that notice it.

**Comment by yreynhout** [comment]
By design / intent, commands have a single destination, events have multiple - that's the basic premise. I think you're abusing command handlers / modules for a purpose they are not fit for, TBH. You're opening a can of worms as to the order in which they will execute in or how errors should be handled, whether or not it makes sense to call the next handler if the previous one failed (assuming sequential, not parallel execution), how this might mess with the transactional boundary if somebody should decide to use it for different reasons. Nothing about this is intuitive to me - sure, you can provide answers and implementations for each of those situations, but then I'd rather code that in the handler module itself.

As an alternative, you can compose the above in the command handling pipeline and tuck most of it away behind a method or enrich handlers as they come out of the module (especially if it is cross cutting and you don't care too much about the particular message).

Anyway, that's just how I solved this in the past - don't let it stop you from going in another direction.

**Comment by CumpsD** [comment]
@yreynhout that's a valid point, about the intent. Forget about this PR :)

**Pull request closed without merging**